streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamFlatTransform.java

```diff
@@ -52,11 +52,12 @@ public void init(final ProcessorContext context) {
         }
 
         @Override
-        public void process(final K1 key, final V1 value) {
-            final KeyValue<? extends K2, ? extends V2> pair = transformer.transform(key, value);
-
-            if (pair != null) {
-                context().forward(pair.key, pair.value);
+        public void process(final KIn key, final VIn value) {
+            final Iterable<KeyValue<KOut, VOut>> pairs = transformer.transform(key, value);
+            if (pairs != null) {
```

Comment:
The JavaDocs says, the return value must not be `null`. With with check, this is actually not true and `null` will be interpreted as empty/zero output records.

I am wondering if we should keep the code as-is and update the JavaDoc, or if we would want to be strict and remove this check and let the for-loop throw an NPE?

\cc @guozhangwang @bbejeck @vvcephei WDYT?

Reply:
Do you refer to the `flatTransform` JavaDocs? There it says
```
     * Method {@link Transformer#transform(Object, Object) transform()} must return an {@link java.util.Iterable} (e.g.,
     * any {@link java.util.Collection} type) and the return value must not be {@code null}.
```
Apparently, I copied this sentence from the `flatMap` JavaDocs. I would remove this sentence from the `flatTransform` JavaDocs because I guess that the only reason to not allow `null`s in a map operation is to distinguish it from a filter operation. This is not needed for a transform operation. Is this correct?

Reply:
Ack.
Please update the JavaDocs accordingly.

For `flatMap()` as it can return zero (ie, empty collection), too, we could also allow to return `null` and interpret it as zero. Don't see a big difference here -- however, for a functional style API it might be ok to be more restrictive and not allow `null`.

Reply:
While I agree that the asymmetry between `flatTransform` and `flatMap` may be confusing, I would also like to note that, we have already a similar asymmetry between `transform` and `map`. The transformer for `transform` has to return `null` to represent the empty result. Consequently, we would also have an asymmetry between `transform` and `flatTransform` if we forbade `null` as a representation of the empty output for `flatTransform`. 

Do we want to have an asymmetry between `flatTransform` and `transform` or between `flatMap` and `flatTransform`?

One option could be to go for the former -- i.e., forbidding `null` output for `flatTransform` -- and to create a JIRA issue to change the transformer for `transform` to return `java.util.Optional<KeyValue<KOut, VOut>>` instead of KeyValue<KOut, VOut> to restore symmetry regarding `null` output values. This solution would break backward compatibility, though.

Reply:
One thing to consider: `map()` is a 1:1 operation while `flatMap()` is a 1:[0,...,n] operation. Also `transform()` in existing API is a 1:[0,...,n] operations. Thus, the grouping and comparison of map/transform and flatMap/flatTransform cannot really be applied.

Note that `map()` and `flatMap()` are DSL operators while `transform()` is PAPI integration. Also note, that `flatMap()` has a different return type compared to `transform()`, that makes it possible to require non-null output for `flatMap()` but not for `transform()` in old Java7 API. 

We introduce `flatTransform()` as new 1:[0,...,n] transformation now and also have Java8. However, as @cadonna mentioned, we cannot simply change the return type for `transform()` to `Optional` though. Also, for PAPI integration I think it's not required to apply the same rules as for DSL operators and to disallow `null` necessarily.

Thus the base question seems to be, should we change the "contract" of `transform() from 1:[0,...,n] to a 1:1 operator because `flatTransform()` is the preferred operation for 1:[0,...,n] now.   I think the answer is _no_, because `transform()` still allows to emit records via `context.forward()`.

Thus, I would keep `map()`, `flatMap()`, and `transform()` as-is, and I would allow `null` as return type for `flatTransform()`, because it's "just" a type safe alternative to `transform()` anyway.

Just my 2 cents.

Reply:
That sounds fine to me @mjsax . I think we can possibly find a better way to unify all this stuff, but it would require some structural changes in the processor context.

I'm happy with permitting null return in `flatTransform`.

Reply:
I removed the above sentence from flatTransform's javadocs.